lib/PromAlertProxy/Config.pm

```diff
@@ -0,0 +1,163 @@
+package PromAlertProxy::Config {
+
+use 5.028;
+use Moo;
+use experimental qw(signatures);
+
+use Types::Standard qw(Str Int HashRef ArrayRef);
+use Type::Utils qw(class_type);
+
+use PromAlertProxy::Logger '$Logger';
+
+use Carp qw(croak);
+use TOML qw(from_toml);
+use Path::Tiny qw(path);
+
+has hub => (
+  is       => 'ro',
+  isa      => class_type('PromAlertProxy::Hub'),
+  required => 1,
+);
+
+has filename => (
+  is       => 'ro',
+  isa      => Str,
+  required => 1,
+);
+
+has _config_raw => (
+  is => 'lazy',
+  isa => HashRef,
+  default => sub ($self) {
+    my ($config, $err) = from_toml(path($self->filename)->slurp);
+    croak "couldn't parse config file '".$self->filename."': $err" unless $config;
+    return $config;
+  },
+);
+
+has _listen_ip => (
+  is  => 'lazy',
+  isa => Str, # XXX IPAddress
+  default => sub ($self) {
+    $self->_config_raw->{server}->{ip};
+  },
+);
+
+has _listen_port => (
+  is  => 'lazy',
+  isa => Int, # XXX Port
+  default => sub ($self) {
+    $self->_config_raw->{server}->{port};
+  },
+);
+
+has _targets => (
+  is => 'lazy',
+  isa => ArrayRef[class_type('PromAlertProxy::Config::Target')],
+  default => sub ($self) {
+    my $config = $self->_config_raw;
+
+    my @targets;
+    for my $id (keys $config->{target}->%*) {
+      my $target_config = $config->{target}->{$id};
+
+      my $target = PromAlertProxy::Config::Target->new(
```

Comment:
In the same breath as "I would like the hub to be extendable at runtime" I also think that `set_hub` sounds terrible because I want things to be fully existing in their final form at instantiation.

If this was C++, `set_hub` would be protected and `Hub` would be a friend and allowed to call it (that would be damaged in a bunch of other different ways).

I'm mostly just musing about the awkwardness of modeling abstract things.

Reply:
You just need to expand your definition of "instantiation," maaaan. (You could totes do it in an `after BUILD` method, which _technically_ be "after instantiation" but not _really_.)